www/common.ts

```diff
@@ -366,28 +366,39 @@ class IonicDeployImpl {
   }
 
   // compare an update to the current version using both name & code
-  private isCurrentVersion(update: IAvailableUpdate) {
+  private isUpdateForCurrentBinary(update: IAvailableUpdate) {
     const currentVersionCode = this._savedPreferences.binaryVersionCode;
     const currentVersionName = this._savedPreferences.binaryVersionName;
     console.log(`Current: versionCode: ${currentVersionCode} versionName: ${currentVersionName}`);
     console.log(`update: versionCode: ${update.binaryVersionCode} versionName: ${update.binaryVersionName}`);
     return update.binaryVersionName === currentVersionName && update.binaryVersionCode === currentVersionCode;
   }
 
-  private async cleanCurrentVersionIfStale() {
+  private isUpdateCurrentlyInstalled(update: IAvailableUpdate) {
+    return this._savedPreferences.currentVersionId === update.versionId;
+  }
+
+  private async cleanupStaleVersions() {
+    const updates = this.getStoredUpdates();
     const prefs = this._savedPreferences;
-    // Is the current version built from a previous binary?
-    if (prefs.currentVersionId) {
-      if (!this.isCurrentVersion(prefs.updates[prefs.currentVersionId]) && !(await this._isRunningVersion(prefs.currentVersionId))) {
+
+    for (const update of updates) {
+      // Is the version built from a previous binary?
+      if (!this.isUpdateForCurrentBinary(update) && !(await this._isRunningVersion(update.versionId))) {
         console.log(
-          `Update ${prefs.currentVersionId} was built for different binary version removing update from device` +
-          `Update binaryVersionName: ${prefs.updates[prefs.currentVersionId].binaryVersionName}, Device binaryVersionName ${prefs.binaryVersionName}` +
-          `Update binaryVersionCode: ${prefs.updates[prefs.currentVersionId].binaryVersionCode}, Device binaryVersionCode ${prefs.binaryVersionCode}`
+          `Update ${update.versionId} was built for different binary version removing update from device` +
+          `Update binaryVersionName: ${update.binaryVersionName}, Device binaryVersionName ${prefs.binaryVersionName}` +
+          `Update binaryVersionCode: ${update.binaryVersionCode}, Device binaryVersionCode ${prefs.binaryVersionCode}`
         );
-        const versionId = prefs.currentVersionId;
-        // NOTE: deleting pref.currentVersionId here to fool deleteVersionById into deleting it
-        delete prefs.currentVersionId;
-        await this.deleteVersionById(versionId);
+
+        await this.deleteVersionById(update.versionId);
+
+        // This is no longer necessary for this function, but a previous version of the code
+        // deleted `prefs.currentVersionId` near initialization so other code may rely on still
+        // deleting `prefs.currentVersionId`.
+        if (this.isUpdateCurrentlyInstalled(update)) {
+          delete prefs.currentVersionId;
```

Comment:
This is now after `deleteVersionById()` which calls `_savePrefs()`... so I think we should keep `delete prefs.currentVersionId;` before `this.deleteVersionById`.